### DISCUSSION TITLE
Improved logging from Modules

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12458] [Ideas]
Questions around how to write progress directly from module code come up fairly often. Right now all modules can do is write to stdout+stderr and it will show up in the progress output in the same way other exec output shows up. Very barebones.

Various problems that have been raised w/ the current situation:
1. Can't differentiate between debug + normal + error logs. You just write every log or no log.
   * In theory you could make the "log level" of your function a parameter, but that creates pretty atrocious DX (passing random parameters everywhere) and in the long term ends up impacting seemingly unrelated things like caching.
2. Lack of support for structured output such as hierarchical display w/ custom labels of function invocations from modules
   * basically, the same sort of idea as the pipeline API
   * related: https://github.com/dagger/dagger/issues/5994

We've discussed various approaches to this. For `1.` in particular we most recently landed on the idea of an actual logging API in graphql that the SDKs would then pick up. Worth more thought and some careful design; it would be nice to have an end result that is compatible with other loggers out there (though trying to do that across languages, which each have their own microcosm of logging frameworks may be a lot to bite off).

For `2.`, it's not clear yet if the existing pipeline API can be fixed and retrofit onto modules or if we need a new approach entirely.

**Comment by shykes** [comment]
Referencing #5156 which is related.

**Comment by shykes** [comment]
My proposal is to focus on *function status* rather than *logs*.

- **Logs** are something you should only look at for deep debugging, auditing, etc. You should *not* look at logs as part of routine day-to-day use of Dagger, including looking at errors ("why did my pipeline fail?"). We need to move away from a logs-centric user experience, both in the CLI and Web. Fundamentally, logs are about showing you a complete history of everything that happened, filtered in various ways depending on what you're looking for. When you need completeness of information, and versatility in how to query it, logs are the right tool. When you need 
- **Status** is something you should be looking at every day, as part of the core Dagger user experience. Did my build succeed? How many functions are running in my pipeline right now, and what are they for? Why did my deploy failed? These are questions that should be answered by rendering *status*, not logs. If I have to read through logs do get an answer, our UI has failed.

One key difference between logs and status: logs are unbounded in size (a function could produce 0 lines, or 10000 likes of logs, impossible to know in advance); whereas the status are bounded (each function is either running or not; either returned an error not; either has a description or not). This makes rendering them to the user much easier.

## Design proposal

Here's a concrete proposal to get bikeshedding started:

- When visualizing a run, the TUI should display no logs by default, whether from exec stdout/stderr (see #5820) or other sources.
- The TUI should render a tree of function calls using the following status information:
  - Function name
  - Function description ("what does this do?")
  - Duration
  - Current status (running, waiting, canceled, completed, failed..)
  - Current task (if running)
  - Error message (if failed)

- Most of these can be set by the engine (and probably already are)
- "Error message" can be harvested by the function return value (insert placeholder for bikeshedding on multiple errors per function, spoiler alert, I think we don't need that)
- That leaves us with "current task" as the only new API call required (so that a function can let the user know "here's what I'm doing"). Insert bikeshedding topic: maybe we don't need that either - maybe if you want to inform the user of a sub-task your function is doing, you split that up into a sub-function, and call it. In that case there is zero new API needed.

I wrote this comment in reaction to a Discord exchange between @sipsma, @nicholasjackson . Would love to get your thoughts! Also @vito since we discussed this in the past.

**Comment by nicholasjackson** [comment]
This looks good to me, I agree that having a clean output is beneficial for normal day to day operation. However, in the instance that if (when, in my case) something goes wrong it is essential to see the output of the command that failed. There are two ways you could do this, and I am not sure that they are both mutually exclusive:

1. run in DEBUG mode, streaming stderr, stdout from the running container, GitHub actions has this as an option after a failed run.
2. Error message is not a basic go error, but a complex one that enables the wrapping of a message, along with `io.Reader` for stderr, and stdout. The TUI could then display the error and the output. I am thinking a common usecase for this would be test failure, the message would be “Tests failed”, the stdout the output from the test command. Error could of course be a string concatenation of message and stdout, but I think a type could be better UX.

 On the subtasks, I think this is useful, for example, my task is Build binaries, the subtask is the architecture. Not really sure if this is necessary though as the task could just be Building arm64. The latter is currently how GitHub and other CIs would manage this. I suppose one advantage of subtask is when you have concurrency, if you had two tasks, each with subtasks all running concurrently, without specific subtask logging the output could get a little messy. I.e.:

```
Building Linux Arm64
Building Darwin Amd64
Packaging Linux Arm64
Building Windows Amd64
```

or

```
Building
 - Linux Arm64
 - Darwin Amd64
 - Windows Amd64
Packaging
 - Linux Amd64
```

**Comment by shykes** [comment]
@nicholasjackson in this scenario, your function should decide what error is presented to the user. If that's the stderr stream of a particular command you execute, then your function would query it via `Container.Stderr()` then return it as an error. If you want to return stdout instead, then you would query and return that. In short, it should be up to the function's developer what to return.

**Comment by nicholasjackson** [comment]
100%; I was thinking more about consistent formatting; if you separate what went wrong from the detail, it will allow you to present a consistent visual treatment. This consistency could be better when consuming errors from modules as the use of color in the output can help immediately draw the eye to the short error message.

i.e.

An un-opinionated approach
```
Run go test -v -race ./...
error: tests failed
testing: warning: no tests to run
PASS
ok  	github.com/jumppad-labs/jumppad	1.116s [no tests to run]
?   	github.com/jumppad-labs/jumppad/cmd	[no test files]
?   	github.com/jumppad-labs/jumppad/cmd/changelog	[no test files]
?   	github.com/jumppad-labs/jumppad/cmd/view	[no test files]
?   	github.com/jumppad-labs/jumppad/examples/build/src	[no test files]

```

An opinionated approach with consistent formatting, `Error` and `StdOut` in this example would be colored, and the stdout log would be indented.
```
Run go test -v -race ./...
**Error**: unit tests failed
**StdOut**:
  testing: warning: no tests to run
  PASS
  ok  	github.com/jumppad-labs/jumppad	1.116s [no tests to run]
  ?   	github.com/jumppad-labs/jumppad/cmd	[no test files]
  ?   	github.com/jumppad-labs/jumppad/cmd/changelog	[no test files]
  ?   	github.com/jumppad-labs/jumppad/cmd/view	[no test files]
  ?   	github.com/jumppad-labs/jumppad/examples/build/src	[no test files]
```

**Comment by jedevc** [comment]
> When visualizing a run, the TUI should display no logs by default, whether from exec stdout/stderr or other sources

I quite like this idea :tada: It would put the emphasis more on the contents of the modules, instead of just on all the data, all the time, and relying on the user to manually filter this.

I do think we need a way to allow a user to surface log data while running a function, that *doesn't* involve restarting it. For example, if I see that a container is taking a long time to run, I would like to be able to maybe just "hit a single key", and I can inspect the contents of those logs and start streaming them to check on what it'd doing.

This is sort of related to the issue we have today where successful steps collapse their logs - so if there was useful information (e.g. a database connection url string), I've now lost it. But I think this would get worse if we collapse all logs by default.

**Comment by vito** [comment]
(Coming in hot!)

tl;dr: 100% yes to higher level representation of failures in general, and status reporting being more useful at a high level especially in the happy path, but I have a ton of doubt around hiding logs in the current state of things, so trying to throw as many wrenches as I can into this to fully suss it out.

> I do think we need a way to allow a user to surface log data while running a function, that _doesn't_ involve restarting it. For example, if I see that a container is taking a long time to run, I would like to be able to maybe just "hit a single key", and I can inspect the contents of those logs and start streaming them to check on what it'd doing.

Totally agree - IMO the live-running DX needs to let me follow the test output as it's happening so I can react to failures in a long integration test suite while letting it to run to completion so I can get a full picture of how many tests failed when I context-switch back. I don't want to sacrifice feedback loop fidelity just to use Dagger.

I don't think this is addressed by "your function should understand the failure and report it" alone - I need _instant_ feedback, not 5-to-10-to-sometimes-20-minute-delayed feedback. I suppose if we take the spirit of that idea and combine it with instant feedback, we would need a way for a function to live-stream individual test results back to the caller. This would be great, and I've often wanted a better TUI for `go test`, but a lot of moving pieces have to work in harmony there (the most worrying being the parsing of `go test` output and being 100% sure it's not missing some critical smoking gun line of output that didn't make it into the right pipe because something unusual happened and I just wasted 2 hours and only found it after running `go test` manually etc etc etc).

If we hide logs until the run completes, or involve too many steps in revealing the logs on each run, I will just not use Dagger as part of my feedback loop and use `go test` instead, kicking me out of my curated portable dev environment. I've [already had to do this for other reasons](https://github.com/dagger/dagger/blob/main/hack/most-tests), would like to avoid adding another.

(Also, total side tangent, but I don't really like the idea of manually grabbing/propagating stdout/stderr for two reasons: 1. they can be ginormous for a runaway build, 2. no one likes troubleshooting a split stdout/stderr output because a human never sees that (we need an API to show the combined output), 3. it seems like pointless extra effort for every module author to implement when we already have all the necessary pieces to show this data _live_)

**Comment by shykes** [comment]
@vito could you address the other side of the tradeoff please? There is a cost to keeping things unchanged for another 6 months, I would like us to discuss that cost.

Last time wrenches were thrown, we just moved on and didn't address the underlying concern (over-reliance on streaming stdout & stderr as the everything-UI).

> it seems like pointless extra effort for every module author to implement when we already have all the necessary pieces to show this data live)

Yes, *all* the data is there, all the time. It's too much data. We're teaching our community that dumping raw logs on our users is good enough. And we're teaching our ecosystem that if they need a more concise output, they should turn to wrappers.

The  default output with `focus=true` hides too much. `focus=false` dumps everything. The flag itself is crucial to know, but cryptic to explain and type.

> successful steps collapse their logs - so if there was useful information (e.g. a database connection url string), I've now lost it

This should no longer be an issue with functions and their return values.

I agree we need good streaming to avoid frustrating users. I just want us to also agree that the current status quo is far from perfect and we need a plan for how to improve it.

**Comment by vito** [comment]
@shykes To be clear, I'm not proposing that we leave things unchanged. I'm trying to figure out how to visualize all this as we speak. :sweat_smile: I'm just hesitant to draw any ecosystem-shaping conclusions from a DX that was never designed for the use case it's now being applied to. Logs and log streaming aren't bad, being unable to interpret them because they're poorly curated and contextualized is bad.

Our progress streaming infra was designed to display Buildkit vertices, which was the right thing to do at one point, but now we have a higher level of abstraction built _on top of_ Buildkit vertices (modules + functions), so there is a _ton_ of extra noise with no way to discern the signal. 

Before Zenith, I don't think it was actually that bad - you would see the commands you ran, and their logs. Pretty straightforward, except it was sometimes hard to tell what the commands were actually doing, hence the addition of the pipelines API, which definitely helped but it never felt like we really nailed that, either.

Keep in mind we didn't even _have_ "focus mode" until the very first Zenith experiments, and it was maybe an hour of thought and work to implement (modulo TUI noodling); it's always been a band-aid.

In essence what I'm proposing is we make a solid effort to visualize Zenith level abstractions now that they're here, using Buildkit data to _enrich_ this visualization rather than overwhelm it, and hopefully retire the pipeline/focus APIs along the way because we just don't need them anymore.

Here's what I'm experimenting with so far, which feels pretty close to that, modulo a bit of extra noise from CLI plumbing:

![image](https://github.com/dagger/dagger/assets/1880/da94e721-0bea-4ff0-b749-a0188ecf95ec)

(...imagining you can cilck the `gotestsum` step to reveal its output, which I haven't implemented yet)

(edit: don't have to imagine anymore)

![image](https://github.com/dagger/dagger/assets/1880/945f39c3-9f25-4aaf-97fe-113604b8f3a3)


Basically, I fully agree that we need to focus on function status first like with the above screenshot, which I tried to call out in the tl;dr. I just don't want to lose log streaming, and I don't want to conclude that logs are evil, because that implies a whole lot of extra work in the ecosystem (albeit minor but widespread), when logs are what every dev is already familiar with, and already has to optimize to some extent (people use these CLIs natively too).

**Comment by shykes** [comment]
This all makes sense to me 👍

**Comment by nipuna-perera** [comment]
Example of cryptic error output - https://github.com/dagger/dagger/issues/6532
